File: python/2025/3/put_marbles_in_bags/put_marbles_in_bags.py
import time
from typing import List

'''
    Reescrita em python do problema desenvolvido em java.
'''


def main() -> None:
    print("=== Testes Put Marbles in Bags ===")
    test_cases = [
        ([1, 3, 5, 1], 2, 4, 1),
        ([1, 3], 2, 0, 2),
        ([1, 3, 5, 6, 2, 8], 2, 7, 2)
    ]

    for weights, k, expected, test_id in test_cases:
        test_case(weights, k, expected, test_id)


def test_case(weights: List[int], k: int, expected: int, test_id: int) -> None:
    print(f"\nTest Case {test_id}:\nInput: {weights}")

    start_time = time.perf_counter_ns()
    result = putMarbles(weights, k)
    elapsed_time = time.perf_counter_ns() - start_time

    print(f"Expected: {expected}")
    print(f"Result:   {result}")
    print(f"Time: {elapsed_time / 1000:.3f} μs")
    print("✅ PASS" if result == expected else "❌ FAIL")


def putMarbles(weights: List[int], k: int) -> int:
    n = len(weights)
    if k == 1 or n == k:
        return 0

    pair_weights = [weights[i] + weights[i + 1] for i in range(n - 1)]
    pair_weights.sort()

    cuts = k - 1
    min_sum = sum(pair_weights[:cuts])
    max_sum = sum(pair_weights[-cuts:]) if cuts > 0 else 0

    return max_sum - min_sum


if __name__ == "__main__":
    main()

File: java/2025/3/putmarblesinbags/PutMarblesInBags.java
import java.util.Arrays;

public class PutMarblesInBags {

    /* Buscamos achar nesse problema a divisão com menor e maior peso.
       Porem o peso da sacola é sempre definido pelo valor do primeiro peso da sacola + peso do ultimo.
       Logo, é possivel encontrar a resposta somando os pares de pesos mais aptos k-1 vezes,
        que representaria a quantidade de cortes feitos no array weights.


        n = weights size
        Time: O(n log n)
        Space (n)
    */

    public static long putMarbles(int[] weights, int k) {
        int n = weights.length;
        if (k == 1 || n == k) return 0;

        int[] pairWeights = new int[n-1];

        // Calcula o peso da divisão em pares para saber o peso de cada corte
        for (int i = 0; i < n-1; i++) {  //                O(n-1)
            pairWeights[i] = weights[i] + weights[i+1]; // * 1
        }

        Arrays.sort(pairWeights); //O(n log n)

        int  cuts = k-1;
        long min = 0;
        long max = 0;

        for (int i = 0; i < cuts; i++) { // O(k-1)
            min += pairWeights[i];  //           * 1
            max += pairWeights[n - 2 - i];  //   * 1
        }

        return max - min;
    }

    // ===== TESTES =====
    public static void main(String[] args) {
        System.out.println("=== Testes Put Marbles in Bags ===");

        testCase(new int[]{1, 3, 5, 1}, 2, 4, 1);
        testCase(new int[]{1, 3}, 2, 0, 2);
        testCase(new int[]{1, 3, 5, 6, 2, 8}, 2, 7, 2);
    }

    private static void testCase(int[] weights, int k, long expected, int testId) {
        long start = System.nanoTime();
        long result = putMarbles(weights, k);
        long duration = (System.nanoTime() - start) / 1000;

        System.out.printf("\nTeste %d:\nInput: %s, k=%d\n", testId, Arrays.toString(weights), k);
        System.out.printf("Resultado: %d (Esperado: %d)\n", result, expected);
        System.out.printf("Tempo: %d μs\n", duration);
        System.out.println(result == expected ? "✅ PASS" : "❌ FAIL");
    }
}